sw/device/tests/crypto/aes_gcm_functest.c

```diff
@@ -0,0 +1,181 @@
+                0x3a22aab7,
+                0x97a0756c,
+                0x79ce3366,
+                0x06dffdd9,
+            },
+        .expected_ciphertext = NULL,
+    },
+
+    // Empty input, empty aad, 16-bit IV, 128-bit key
+    {
+        .key_len = kAesKeyLen128,
+        .iv_len = 16,
+        .iv =
+            {// IV = 22294cae82d82e44427dfcc33bacdbec
+             0xae4c2922, 0x442ed882, 0xc3fc7d42, 0xecdbac3b},
+        .plaintext_len = 0,
+        .plaintext = NULL,
+        .aad_len = 0,
+        .aad = NULL,
+        .expected_tag =
+            {// Tag = 4c59f0d420d9eb8669c40ad23b5419ba
+             0xd4f0594c, 0x86ebd920, 0xd20ac469, 0xba19543b},
+        .expected_ciphertext = NULL,
+    },
+
+    // 16-bit IV, 256-bit key, real message and aad
+    {
+        .key_len = kAesKeyLen256,
+        .iv_len = 16,
+        .iv =
+            {// IV = c58aded2e1bbecba8b16a5757e5475bd
+             0xd2de8ac5, 0xbaecbbe1, 0x75a5168b, 0xbd75547e},
+        .plaintext_len = kPlaintextLen,
+        .plaintext = kPlaintext,
+        .aad_len = kAadLen,
+        .aad = kAad,
+        .expected_tag =
+            {// Tag = 324895b3d2f656e4fa2f8ce056137061
+             0xb3954832, 0xe456f6d2, 0xe08c2ffa, 0x61701356},
+        .expected_ciphertext = kCiphertext256,
+    },
+};
+
+OTTF_DEFINE_TEST_CONFIG();
+bool test_main(void) {
+  for (size_t i = 0; i < ARRAYSIZE(aes_gcm_testvectors); i++) {
+    LOG_INFO("Starting AES-GCM test %d of %d...", i + 1,
+             ARRAYSIZE(aes_gcm_testvectors));
+    aes_gcm_test_t test = aes_gcm_testvectors[i];
+
+    // Construct key shares by setting first share to full key and second share
+    // to 0. (Note: this is not a secure construction! But it makes debugging
+    // tests easier because there is only one thing to print.)
+    uint32_t *key_shares[2];
+    uint32_t share1[8] = {0};
+    key_shares[1] = share1;
+    if (test.key_len == kAesKeyLen128) {
+      key_shares[0] = kKey128;
+    } else if (test.key_len == kAesKeyLen256) {
+      key_shares[0] = kKey256;
+    } else {
+      LOG_ERROR("No key available for key length.");
+      return false;
+    }
+
+    // Call AES-GCM encrypt.
+    uint8_t actual_ciphertext[test.plaintext_len];
+    uint8_t actual_tag[16];
+    uint64_t start = ibex_mcycle_read();
+    aes_error_t err = aes_gcm_encrypt(
+        test.key_len, key_shares, test.iv_len, (uint8_t *)test.iv,
+        test.plaintext_len, (uint8_t *)test.plaintext, test.aad_len,
+        (uint8_t *)test.aad, actual_ciphertext, actual_tag);
+    uint64_t end = ibex_mcycle_read();
+    uint32_t cycles = end - start;
+    LOG_INFO("aes_gcm_encrypt() took %u cycles", cycles);
+    CHECK(err == kAesOk, "AES-GCM encryption returned an error: %08x", err);
+
+    CHECK_ARRAYS_EQ((uint32_t *)actual_tag, test.expected_tag, 4);
+    if (test.plaintext_len > 0) {
+      CHECK_ARRAYS_EQ((uint32_t *)actual_ciphertext, test.expected_ciphertext,
```

Comment:
Casting `uint8_t*` to `uint32_t*` might violate the strict aliasing assumption. It's always safe to cast to `char*`, so maybe you could do `CHECK_ARRAYS_EQ((const char*)actual_ciphertext, (const char*)test.expected_ciphertext, test.plaintext_len)` here?

Reply:
I get errors when I try to use `char *` or `const char *`.  Looks like under the hood of `CHECK_ARRAYS_EQ` it tries to plug `char` into a `Generic` and that causes issues:
```
sw/device/tests/crypto/aes_gcm_functest.c:171:5: error: controlling expression type 'char' not compatible with any generic association type
    CHECK_ARRAYS_EQ((const char *)actual_tag, (const char *)test.expected_tag,
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sw/device/lib/testing/test_framework/check.h:109:19: note: expanded from macro 'CHECK_ARRAYS_EQ'
        LOG_ERROR(SHOW_MISMATCH_FMT_STR_((actual_)[i]), i, (actual_)[i],      \
        ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./sw/device/lib/testing/test_framework/check.h:58:44: note: expanded from macro 'SHOW_MISMATCH_FMT_STR_'
#define SHOW_MISMATCH_FMT_STR_(a) _Generic((a),                 \
```

To avoid nasty workarounds I've just replaced all the tests' expected values with byte arrays instead of word arrays so everything matches, which allowed me to remove all the casts. This makes it slightly more annoying to enter test data but seems preferable to e.g. inserting `memcpy()` -- and in the long run test data will probably be formatted by an automated script anyway.

Reply:
Oh fun, I was not familiar with `_Generic`. I think you could get around it by adding a case for the `char` type, but I fully support the byte arrays workaround as well.

```diff
--- a/sw/device/lib/testing/test_framework/check.h
+++ b/sw/device/lib/testing/test_framework/check.h
@@ -57,6 +57,7 @@
 // clang-format off
 #define SHOW_MISMATCH_FMT_STR_(a) _Generic((a),                 \
           bool: "CHECK-fail: [%d] got: 0x02%x; want: 0x02%x",   \
+          char: "CHECK-fail: [%d] got: 0x02%x; want: 0x02%x",   \
         int8_t: "CHECK-fail: [%d] got: 0x02%x; want: 0x02%x",   \
        uint8_t: "CHECK-fail: [%d] got: 0x02%x; want: 0x02%x",   \
        int16_t: "CHECK-fail: [%d] got: 0x04%x; want: 0x04%x",   \
@@ -67,6 +68,7 @@
       uint64_t: "CHECK-fail: [%d] got: 0x016%x; want: 0x016%x")
 #define SHOW_MATCH_FMT_STR_(a) _Generic((a),            \
           bool: "CHECK-fail: [%d] both equal: 0x02%x",  \
+          char: "CHECK-fail: [%d] both equal: 0x02%x",  \
         int8_t: "CHECK-fail: [%d] both equal: 0x02%x",  \
        uint8_t: "CHECK-fail: [%d] both equal: 0x02%x",  \
        int16_t: "CHECK-fail: [%d] both equal: 0x04%x",  \
```